common/models/role.js

```diff
@@ -244,31 +256,39 @@ module.exports = function(Role) {
         return;
       }
       debug('Model found: %j', inst);
+
+      // loopback v2 implementation alows to resolve isOwner() if principalType is USER,
+      // instance ownerId (.userId or .owner) exists and is equal to user's id
       var ownerId = inst.userId || inst.owner;
-      // Ensure ownerId exists and is not a function/relation
-      if (ownerId && 'function' !== typeof ownerId) {
-        if (callback) callback(null, matches(ownerId, userId));
+      if (principalType === Principal.USER && ownerId && 'function' !== typeof ownerId) {
+        callback(null, matches(ownerId, userId));
         return;
       } else {
         // Try to follow belongsTo
         for (var r in modelClass.relations) {
           var rel = modelClass.relations[r];
-          if (rel.type === 'belongsTo' && isUserClass(rel.modelTo)) {
-            debug('Checking relation %s to %s: %j', r, rel.modelTo.modelName, rel);
+          // relation should be belongsTo and target a User based class
+          if (rel.type !== 'belongsTo' && !isUserClass(rel.modelTo)) {
+            continue;
+          }
+          // checking related user
+          var userModelName = rel.modelTo.modelName;
+          if (principalType === Principal.USER || principalType === userModelName) {
+            debug('Checking relation %s to %s: %j', r, userModelName, rel);
             inst[r](processRelatedUser);
             return;
           }
```

Comment:
we now need this additional check
`if (principalType === Principal.USER || principalType === userModelName) {...`
 to make sure we don't follow a belongTo relation towards an incorrect user model.

**ps**. in order to keep the level of indent controlled, i chose to first `continue` if `rel.type !== 'belongsTo' && !isUserClass(rel.modelTo)`